heat/core/signal.py

```diff
@@ -0,0 +1,147 @@
+"""Provides a collection of signal-processing operations"""
+
+import torch
+from typing import Union, Tuple, Sequence
+
+from .communication import MPI
+from .dndarray import DNDarray
+from .types import promote_types
+from .manipulations import pad
+from .factories import array
+import torch.nn.functional as fc
+
+__all__ = ["convolve"]
+
+
+def convolve(a: DNDarray, v: DNDarray, mode: str = "full") -> DNDarray:
+    """
+    Returns the discrete, linear convolution of two one-dimensional `DNDarray`s.
+
+    Parameters
+    ----------
+    a : DNDarray
+        One-dimensional signal `DNDarray` of shape (N,)
+    v : DNDarray
+        One-dimensional filter weight `DNDarray` of shape (M,).
+    mode : str
+        Can be 'full', 'valid', or 'same'. Default is 'full'.
+        'full':
+          Returns the convolution at
+          each point of overlap, with an output shape of (N+M-1,). At
+          the end-points of the convolution, the signals do not overlap
+          completely, and boundary effects may be seen.
+        'same':
+          Mode 'same' returns output  of length 'N'. Boundary
+          effects are still visible. This mode is not supported for
+          even-sized filter weights
+        'valid':
+          Mode 'valid' returns output of length 'N-M+1'. The
+          convolution product is only given for points where the signals
+          overlap completely. Values outside the signal boundary have no
+          effect.
+
+    Notes
+    -----
+        Contrary to the original `numpy.convolve`, this function does not
+        swap the input arrays if the second one is larger than the first one.
+        This is because `a`, the signal, might be memory-distributed,
+        whereas the filter `v` is assumed to be non-distributed,
+        i.e. a copy of `v` will reside on each process.
+
+
+    Examples
+    --------
+    Note how the convolution operator flips the second array
+    before "sliding" the two across one another:
+    >>> a = ht.ones(10)
+    >>> v = ht.arange(3).astype(ht.float)
+    >>> ht.convolve(a, v, mode='full')
+    DNDarray([0., 1., 3., 3., 3., 3., 2.])
+    >>> ht.convolve(a, v, mode='same')
+    DNDarray([1., 3., 3., 3., 3.])
+    >>> ht.convolve(a, v, mode='valid')
+    DNDarray([3., 3., 3.])
```

Comment:
```suggestion
    Note how the convolution operator flips the second array
    before "sliding" the two across one another:
    
    >>> a = ht.ones(10)
    >>> v = ht.arange(3).astype(ht.float)
    >>> ht.convolve(a, v, mode='full')
    DNDarray([0., 1., 3., 3., 3., 3., 2.])
    >>> ht.convolve(a, v, mode='same')
    DNDarray([1., 3., 3., 3., 3.])
    >>> ht.convolve(a, v, mode='valid')
    DNDarray([3., 3., 3.])
```

Reply:
I thought we were not supposed to have blank lines within the examples section. By all means go ahead and make the change, thanks!

Reply:
The docstring is rendered poorly when there is no blank line. Alternatively, you can use `\n` instead.